compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -3293,27 +3345,49 @@ public Boolean visit(BParameterizedType t, BType s) {
         }
     }
 
-    private boolean checkUnionHasSameType(LinkedHashSet<BType> memberTypes, BType baseType) {
-        boolean isSameType = false;
+    private boolean isNil(BType type) {
+        // Currently, type reference for `null` literal is taken as Finite type and type reference for `()` literal
+        // taken as Nil type.
+        BType referredType = getReferredType(type);
+        TypeKind referredTypeKind = referredType.getKind();
+        if (referredTypeKind == TypeKind.NIL) {
+            return true;
+        } else if (referredTypeKind == TypeKind.FINITE) {
+            Set<BLangExpression> valueSpace = ((BFiniteType) referredType).getValueSpace();
+            return valueSpace.size() == 1 && valueSpace.iterator().next().getBType().getKind() == TypeKind.NIL;
+        }
+        return false;
+    }
+
+    private boolean checkUnionHasSameType(BUnionType unionType, BType baseType) {
+        LinkedHashSet<BType> memberTypes = unionType.getMemberTypes();
         for (BType type : memberTypes) {
             type = getReferredType(type);
             if (type.tag == TypeTags.FINITE) {
                 for (BLangExpression expr : ((BFiniteType) type).getValueSpace()) {
-                    isSameType = isSameOrderedType(expr.getBType(), baseType);
-                    if (!isSameType) {
+                    if (!isSameOrderedType(expr.getBType(), baseType)) {
                         return false;
                     }
                 }
             } else if (type.tag == TypeTags.UNION) {
```

Comment:
This scenario is not covered with unit tests (locally test with commenting out this). I guess, unions cannot be a member type of another union. Therefore, we cannot cover this part with tests. IMO we can remove this condition. But this condition is already available in the master branch. WDYT @KavinduZoysa?

Reply:
Please check whether tests are passing after removing this test.

Reply:
Tests are passing even without this. (https://github.com/ballerina-platform/ballerina-lang/pull/37047/commits/85848bce6e929fd06d4002aeec1f9351f6ec1634)

Reply:
We can go head.